### PR TITLE
next-mdx-remote v3

### PR DESCRIPTION
**Pull request by BRKalow**
Rewrite of `next-mdx-remote` to fix a number of longstanding issues, simplify usage, and improve performance!

Credit to @lfades for his work in https://github.com/hashicorp/next-mdx-remote/pull/39, which laid the groundwork for this.

To install:
```
npm install next-mdx-remote@next
```

Migration guide: https://www.npmjs.com/package/next-mdx-remote/v/3.0.0-alpha.0#migrating-to-v3

Feel free to leave feedback here if you test it out!

### CODE REVIEW

**Comment by artemtam** [comment]
Hey! Had to test it out, because version 2 did not work with my Next.js/TypeScript project. Didn't face any issues yet (checked a single rich article).

Do you need any help with version 3? Could not find any roadmap/issues you are including in the version to work on.

**Comment by BRKalow** [comment]
@artemtam Thanks for testing it out! We're doing some validation on our internal applications, but for the most part this is feature complete. Letting us know that this works for you is great help! 🙏

**Comment by anishathalye** [comment]
I got a chance to try this out, and it solves my issues brought up in #118 / $120. Migration was painless, following the migration guide.

**Comment by payapula** [comment]
Awesome ❤️  

> This should also fix a number of SSR-related CSS-in-JS bugs users were experiencing

This solves the SSR issue with Chakra UI mentioned in - https://github.com/hashicorp/next-mdx-remote/issues/72

I tested it out and the problem seems to be resolved and I can clearly notice improvements in page loads. 

Thanks 👍🏼

**Comment by joshwcomeau** [comment]
Hey there!

I gave the migration a shot, and for the most part, it works great. Was able to remove some SSR hacks for styled-components, and it feels nice and fast =)

One issue, though: The new `MDXProvider` component seems to wrap the rendered content in a `<div>`. I found it in the PR, here: https://github.com/hashicorp/next-mdx-remote/pull/121/files#diff-0b5adbfe7b36e4ae2f479291e20152e33e940f7f265162d77f40f6bdb5da7405R99

This `div` causes problems, since I'm currently rendering each MDX element as a row in a CSS grid.

Might it be possible to switch it to a React fragment, or remove it altogether?

Thanks so much for your work on this!

**Comment by BRKalow** [comment]
@joshwcomeau @payapula Thanks for taking the time to try this out and leave feedback! 🙏 

I've removed the wrapping `div` by default, feel free to pull down `next-mdx-remote@next` again if you get a chance!

**Comment by leerob** [comment]
The latest version is working for me!

File: src/index.tsx [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+  lazy,
+}: MDXRemoteProps) {
+  const [isReadyToRender, setIsReadyToRender] = useState(
+    !lazy || typeof window === 'undefined'
+  )
+
+  // if we're on the client side, we hydrate the mdx content inside
+  // requestIdleCallback, since we can be fairly confident that
+  // markdown - embedded components are not a high priority to get
+  // to interactive compared to...anything else on the page.
```
Comment (jescalan) [inline src/index.tsx:58]:
```suggestion
  // if we're on the client side and `lazy` is set to true, we hydrate the
  // mdx content inside requestIdleCallback, allowing the page to get to
  // interactive quicker, but the mdx content to hydrate slower.
```

**Review by jescalan — APPROVED** [review]
👏

**Comment by joshwcomeau** [comment]
@BRKalow Woohoo! Thanks for this :)

I did notice one edge-case regression.

In my MDX, I had a component like this:

```mdx
<MultipleChoiceQuestion
  question={
    <>
      Something something
    </>
  }
/>
```

When the `renderPage` method is called, I get an error, `React is not defined`. I figured out that it was because of the fragment, `<>`. In 2.x, this works fine, but in 3.0 it doesn't seem to make React available directly, when parsing the elements in MDX.

I worked around this by switching `<>` to `<Fragment>`, and updating my `components` map to include it:

```jsx
<MDXRemote 
  components={{
    Fragment: React.Fragment,
    // All my other stuff
  }}
/>
```

Let me know if you'd like me to open an issue with this info!

**Comment by BRKalow** [comment]
😱 oh no! If you could open an issue that would be great! I'll take a look right now.

**Comment by joshwcomeau** [comment]
> 😱 oh no! If you could open an issue that would be great! I'll take a look right now.

Done! https://github.com/hashicorp/next-mdx-remote/issues/127

**Pull request merged**